### PR TITLE
daemon: spool sideloaded snap into blob dir

### PR DESCRIPTION
**Pull request by chipaca**
Before this change a sideloaded snap would get written out to /tmp
before being copied into /var/lib/snapd/snaps. As these are often
separate devices this means sideloading a snap would be unnecessarily
slow, and use up a potentially large amount of /tmp (which might be
small and in-memory, on core).

This change makes it so that the snap is instead written to a
temporary (and hidden) file in /var/lib/snapd/snaps directly.

This fixes lp:1667865.

### CODE REVIEW

File: daemon/api.go [hunk 1]
```diff
@@ -1489,7 +1489,7 @@ out:
 	// we are in charge of the tempfile life cycle until we hand it off to the change
 	changeTriggered := false
 	// if you change this prefix, look for it in the tests
-	tmpf, err := ioutil.TempFile("", "snapd-sideload-pkg-")
+	tmpf, err := ioutil.TempFile(dirs.SnapBlobDir, ".snapd-sideload-")
```
Comment (mvo5) [inline daemon/api.go]: Looks great! Do we need to ensure that we have something that cleans /var/lib/snapd/snaps/.snapd-sideload-* on bootup or periodically when e.g. power fails while such an operation is in progress?

Comment (chipaca) [reply daemon/api.go]: snapd itself could do it on startup

File: overlord/snapstate/snapmgr.go [hunk 2]
```diff
@@ -575,6 +578,51 @@ func (m *SnapManager) atSeed() error {
...
 
+var sideloadCleanupWait = time.Duration(24 * time.Hour)
+
+// cleanOldTemps removes files that might've been left behind by an
+// old aborted sideload
+//
+// They're usually cleaned up, but if they're created and then snapd
+// stops before writing the change to disk (killed, light cut, etc)
+// it'll be left behind.
+func (m *SnapManager) cleanOldTemps() error {
```
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/snapstate/snapmgr.go]: This is going to be called very often while changes are in progress, it should probably have a check that is not run more often than every 10mins or longer.

File: osutil/unlink.go [hunk 3]
```diff
@@ -59,3 +64,9 @@ func UnlinkMany(dirname string, filenames []string) error {
 	}
 	return nil
 }
+
+// FUnlinkMany is like UnlinkMany but takes an open directory *os.File
+// instead of a dirname.
+func FUnlinkMany(dir *os.File, filenames []string) error {
```
[the remark this answers is not in the document]
Comment (chipaca) [reply osutil/unlink.go]: there are a bunch in  `osutil/sys` where the F means a file descriptor,  but nothing with F being an `*os.File`. If `UnlinkManyAt` is clearer, I'm all for it.

**Pull request merged**